// loopback4-soft-delete/dist/models/soft-delete-entity.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SoftDeleteEntity = void 0;
const tslib_1 = require("tslib");
const repository_1 = require("@loopback/repository");
/**
 * Abstract base class for all soft-delete enabled models
 *
 * @description
 * Base class for all soft-delete enabled models created.
 * It adds three attributes to the model class for handling soft-delete,
 * namely, 'deleted', deletedOn, deletedBy
 * Its an abstract class so no repository class should be based on this.
 */
class SoftDeleteEntity extends repository_1.Entity {
    constructor(data) {
        super(data);
    }
}
tslib_1.__decorate([
    (0, repository_1.property)({
        type: 'boolean',
        default: false,
    }),
    tslib_1.__metadata("design:type", Boolean)
], SoftDeleteEntity.prototype, "deleted", void 0);
tslib_1.__decorate([
    (0, repository_1.property)({
        type: 'date',
        name: 'deleted_on',
        jsonSchema: {
            nullable: true,
        },
    }),
    tslib_1.__metadata("design:type", Date)
], SoftDeleteEntity.prototype, "deletedOn", void 0);
tslib_1.__decorate([
    (0, repository_1.property)({
        type: 'string',
        name: 'deleted_by',
        jsonSchema: {
            nullable: true,
        },
    }),
    tslib_1.__metadata("design:type", String)
], SoftDeleteEntity.prototype, "deletedBy", void 0);
exports.SoftDeleteEntity = SoftDeleteEntity;
