// loopback4-soft-delete/dist/mixins/soft-delete-entity.mixin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SoftDeleteEntityMixin = void 0;
const tslib_1 = require("tslib");
const repository_1 = require("@loopback/repository");
function SoftDeleteEntityMixin(base, config) {
    var _a, _b, _c;
    class SoftDeleteEntity extends base {
    }
    tslib_1.__decorate([
        (0, repository_1.property)({
            type: 'boolean',
            default: false,
            ...((_a = config === null || config === void 0 ? void 0 : config.deleted) !== null && _a !== void 0 ? _a : {}),
        }),
        tslib_1.__metadata("design:type", Boolean)
    ], SoftDeleteEntity.prototype, "deleted", void 0);
    tslib_1.__decorate([
        (0, repository_1.property)({
            type: 'date',
            name: 'deleted_on',
            jsonSchema: {
                nullable: true,
            },
            ...((_b = config === null || config === void 0 ? void 0 : config.deletedOn) !== null && _b !== void 0 ? _b : {}),
        }),
        tslib_1.__metadata("design:type", Date)
    ], SoftDeleteEntity.prototype, "deletedOn", void 0);
    tslib_1.__decorate([
        (0, repository_1.property)({
            type: 'string',
            name: 'deleted_by',
            jsonSchema: {
                nullable: true,
            },
            ...((_c = config === null || config === void 0 ? void 0 : config.deletedBy) !== null && _c !== void 0 ? _c : {}),
        }),
        tslib_1.__metadata("design:type", String)
    ], SoftDeleteEntity.prototype, "deletedBy", void 0);
    return SoftDeleteEntity;
}
exports.SoftDeleteEntityMixin = SoftDeleteEntityMixin;
